### DISCUSSION TITLE
Branching on checked vs unchecked item

### DISCUSSION THREAD

**Discussion by lewisl9029** [ariakit/ariakit#1353] [Questions] [answered]
Hi there!

I looked into Select components from a few of the headless libs out there and the one in ariakit has been my favorite by far. Thank you so much for working on it!

One small thing that felt a little bit clunky was when I was implementing the checkboxes in this multiselect design:

![Screen Shot 2022-05-13 at 2 47 43 PM](https://user-images.githubusercontent.com/6934200/168393889-27517798-42d2-450d-b0f4-8609999d5ce1.png)

I ended up having to create a wrapper around the SelectItemCheck component with a `checked={state.value.includes(category.id)}` prop to implement this design. This has been preventing me from abstracting away the SelectItem part of the implementation into its own self-contained reusable component that's only responsible for rendering the current item (since it now needs to have access to all items).

I think a more ergonomic API could involve exposing a render prop from SelectItemCheck with its internal `checked` prop (as well as the checkbox that it renders by default as `check`?), and/or possibly exposing the underlying context through a `useSelectItemChecked` hook. Thoughts?

Some relevant code snippets:

Implementation:
```js
const SelectValue = ({ values }) => {
  if (values.length === 0) {
    return <Body4>API Categories</Body4>;
  }
  return (
    <SelectValueLayout>
      <Body4>API Categories</Body4>
      <Badge>{values.length}</Badge>
    </SelectValueLayout>
  );
};

const SelectMultiple = ({ values, onChange, children }) => {
  const state = ariakitSelect.useSelectState({
    value: values,
    setValue: onChange,
    sameWidth: true,
    gutter: 16,
  });

  return (
    <div>
      <Select state={state}>
        <SelectValue values={state.value} />
        <SelectArrow />
      </Select>
      {state.mounted && <SelectPopover state={state}>{children}</SelectPopover>}
    </div>
  );
};
```

Usage:

```js
        <SelectMultiple
          values={categoriesSelected}
          onChange={setCategoriesSelected}
        >
          {categories.map(category => (
            // TODO: abstract this away into a reusable component, currently blocked by
            // https://github.com/ariakit/ariakit/discussions/1353
            // Unless we add our own context
            <SelectItem key={category.id} value={category.id}>
              <SelectItemCheckContainer
                checked={categoriesSelected.includes(category.id)}
              >
                <SelectItemCheck />
              </SelectItemCheckContainer>
              <Body4>{category.label}</Body4>
            </SelectItem>
          ))}
        </SelectMultiple>
```

Update: also realized there might be a somewhat similar limitation with determining the current active item for branching styles on keyboard navigation. The `data-active-item` attribute works fine for targeting using selectors, but I'm using CSS-in-JS and would prefer to have a prop to branch on instead to keep the vdom as the source of truth to avoid potential race conditions in concurrent mode (currently I've been falling back to using a nested selector).

**Comment by diegohaz** [comment, marked as answer]
You can achieve it using `SelectItem` render props:

```jsx
<SelectItem>
  {(props) => {
    // "data-active-item" in props
    // props["aria-selected"]
  }}
</SelectItem>
```

We still need docs stating which props are safe to use as selectors, but those two are fine.

**Reply by diegohaz** [reply]
> currently I've been falling back to using a nested selector

That's the approach I'd use, btw.

**Reply by lewisl9029** [reply]
Awesome, the render props on SelectItem work great!

Here's what the updated code looks like:

Implementation:
```js
const SelectSummary = ({ summary, length }) => {
  if (length === 0) {
    return <Body4>{summary}</Body4>;
  }
  return (
    <SelectSummaryLayout>
      <Body4>{summary}</Body4>
      <Badge>{length}</Badge>
    </SelectSummaryLayout>
  );
};

const SelectMultiple = ({ values, onChange, summary, children }) => {
  const state = ariakitSelect.useSelectState({
    value: values,
    setValue: onChange,
    sameWidth: true,
    gutter: 16,
  });

  return (
    <div>
      <Select state={state}>
        <SelectSummary summary={summary} length={state.value.length} />
        <SelectArrow />
      </Select>
      {state.mounted && <SelectPopover state={state}>{children}</SelectPopover>}
    </div>
  );
};

const SelectItem = ({ id, label }) => {
  return (
    <ariakitSelect.SelectItem key={id} value={id}>
      {props => (
        <SelectItemCustom {...props} active={props['data-active-item'] === ''}>
          <SelectItemCheckContainer checked={props['aria-selected']}>
            <SelectItemCheck />
          </SelectItemCheckContainer>
          <Body4>{label}</Body4>
        </SelectItemCustom>
      )}
    </ariakitSelect.SelectItem>
  );
};
```

```js
        <SelectMultiple
          summary="API Categories"
          values={categoriesSelected}
          onChange={setCategoriesSelected}
        >
          {categories.map(category => (
            <SelectItem id={category.id} label={category.label} />
          ))}
        </SelectMultiple>
```

One thing that threw me off for a moment was the fact that `data-active-item`'s value was an empty string, which is falsy, compared to `aria-selected` which was a proper boolean (`true`).

**Reply by diegohaz** [reply]
You can check `data-*` attributes that can be used as selectors with `"data-active-item" in props`.

**Reply by emillaine** [reply]
Hey @diegohaz! How can this be solved today? When using the solution suggested above I get "The `children` prop as a function is deprecated. Use the `render` prop instead". When using the `render` prop there's no `aria-selected` passed to it.

**Reply by diegohaz** [reply]
> Hey @diegohaz! How can this be solved today? When using the solution suggested above I get "The `children` prop as a function is deprecated. Use the `render` prop instead". When using the `render` prop there's no `aria-selected` passed to it.

Could you open a new issue? You should have access to the same props when using the render prop. The only difference should be that with the render prop you also receive the children, if any.
